core/src/main/java/org/elasticsearch/common/util/NamedEnum.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.util;
+
+import java.util.EnumSet;
+import java.util.Objects;
+
+/**
+ * An interface that enums should implement if the elements in the
+ * enumeration have different values than their
+ */
+public interface NamedEnum {
+
+    /**
+     * The name by which the enumeration is identified.
+     * @return The name of the enum
+     */
+    String getName();
+
+    /**
+     * Get the correct enum value for the given enum type and name.  The enum
+     * type must implement the {@link NamedEnum} interface for this to work.
+     *
+     * @param cls The class of the enum
+     * @param name The name of the enum within the given type
+     * @return The enum value matching the given name
+     * @throws IllegalArgumentException if the input is invalid or an enum value with the given name could not be found
+     */
```

Comment:
Plus I just realized that sometimes we have more than one variant of a String in the query DSL (e.g. camel casing etcll) that needs to be mapped to one enum. For example, the StringDistance currently allows both `"damerau_levenshtein"` and  `"damerauLevenshtein"`. This might be something to get rid of, but in the meantime we need to support it I think.